src/src/PluginStructs/P128_data_struct.cpp

```diff
@@ -226,113 +229,101 @@ bool P128_data_struct::plugin_write(struct EventStruct *event,
     }
 
     else if (subCommand == F("hsvone")) {
-      mode = On;
+      mode = P128_modetype::On;
       rgb  =
-        RgbColor(HsbColor(parseString(string, 4).toFloat() / 360, parseString(string, 5).toFloat() / 100,
-                          parseString(string, 6).toFloat() / 100));
+        RgbColor(HsbColor(str4.toFloat() / 360, str5.toFloat() / 100,
+                          str6.toFloat() / 100));
 
-      colorStr               = "";
-      rgb.R < 16 ? colorStr  = "0" : "";
-      colorStr              += formatToHex(rgb.R, {});
-      rgb.G < 16 ? colorStr += "0" : "";
-      colorStr              += formatToHex(rgb.G, {});
-      rgb.B < 16 ? colorStr += "0" : "";
-      colorStr              += formatToHex(rgb.B, {});
+      rgb2colorStr();
 
       hex2rgb(colorStr);
-      uint16_t pixnum = parseString(string, 3).toInt() - 1;
+      uint16_t pixnum = str3i - 1;
       Plugin_128_pixels->SetPixelColor(pixnum, rgb);
     }
 
     else if (subCommand == F("hsvline")) {
-      mode = On;
+      mode = P128_modetype::On;
 
       rgb =
-        RgbColor(HsbColor(parseString(string, 5).toFloat() / 360, parseString(string, 6).toFloat() / 100,
-                          parseString(string, 7).toFloat() / 100));
+        RgbColor(HsbColor(str5.toFloat() / 360, str6.toFloat() / 100,
+                          str7.toFloat() / 100));
 
-      colorStr               = "";
-      rgb.R < 16 ? colorStr  = "0" : "";
-      colorStr              += formatToHex(rgb.R, {});
-      rgb.G < 16 ? colorStr += "0" : "";
-      colorStr              += formatToHex(rgb.G, {});
-      rgb.B < 16 ? colorStr += "0" : "";
-      colorStr              += formatToHex(rgb.B, {});
+      rgb2colorStr();
 
       hex2rgb(colorStr);
 
-      for (int i = 0; i <= (parseString(string, 4).toInt() - parseString(string, 3).toInt() + pixelCount) % pixelCount; i++) {
-        Plugin_128_pixels->SetPixelColor((i + parseString(string, 3).toInt() - 1) % pixelCount, rgb);
+      for (int i = 0; i <= (str4i - str3i + pixelCount) % pixelCount; i++) {
```

Comment:
A for loop with the invariant containing a modulo always seems tricky to me as the loop may be shorter than you expect.

Reply:
Again, existing code, I tried not to mess with the algorithms.